tarantool/msgpack_ext/types/datetime.py

```diff
@@ -0,0 +1,260 @@
+                                        nanosecond=nanosecond)
+
+        if tz != '':
+            if tz not in tt_timezones.timezoneToIndex:
+                raise ValueError(f'Unknown Tarantool timezone "{tz}"')
+
+            tzinfo = get_python_tzinfo(tz, ValueError)
+            self._datetime = datetime.replace(tzinfo=tzinfo)
+            self._tz = tz
+        elif tzoffset != 0:
+            tzinfo = pytz.FixedOffset(tzoffset)
+            self._datetime = datetime.replace(tzinfo=tzinfo)
+            self._tz = ''
+        else:
+            self._datetime = datetime
+            self._tz = ''
+
+    def __eq__(self, other):
+        if isinstance(other, Datetime):
+            return self._datetime == other._datetime
+        elif isinstance(other, pandas.Timestamp):
+            return self._datetime == other
+        else:
+            return False
+
+    def __str__(self):
+        return self._datetime.__str__()
+
+    def __repr__(self):
+        return f'datetime: {self._datetime.__repr__()}, tz: "{self.tz}"'
+
+    def __copy__(self):
+        cls = self.__class__
+        result = cls.__new__(cls)
+        result.__dict__.update(self.__dict__)
+        return result
+
+    def __deepcopy__(self, memo):
+        cls = self.__class__
+        result = cls.__new__(cls)
+        memo[id(self)] = result
+        for k, v in self.__dict__.items():
+            setattr(result, k, deepcopy(v, memo))
+        return result
+
+    @property
+    def year(self):
+        return self._datetime.year
+
+    @property
+    def month(self):
+        return self._datetime.month
+
+    @property
+    def day(self):
+        return self._datetime.day
+
+    @property
+    def hour(self):
+        return self._datetime.hour
+
+    @property
+    def minute(self):
+        return self._datetime.minute
+
+    @property
+    def sec(self):
+        return self._datetime.second
+
+    @property
+    def nsec(self):
+        # microseconds + nanoseconds
+        return self._datetime.value % NSEC_IN_SEC
+
+    @property
+    def timestamp(self):
+        return self._datetime.timestamp()
+
+    @property
+    def tzoffset(self):
+        if self._datetime.tzinfo is not None:
+            return compute_offset(self._datetime)
+        return 0
+
+    @property
+    def tz(self):
+        return self._tz
+
+    def msgpack_encode(self):
+        seconds = self._datetime.value // NSEC_IN_SEC
+        nsec = self.nsec
+        tzoffset = self.tzoffset
+
+        tz = self.tz
+        if tz != '':
+            tzindex = tt_timezones.timezoneToIndex[tz]
+        else:
+            tzindex = 0
+
+        buf = get_int_as_bytes(seconds, SECONDS_SIZE_BYTES)
+
+        if (nsec != 0) or (tzoffset != 0) or (tzindex != 0):
+            buf = buf + get_int_as_bytes(nsec, NSEC_SIZE_BYTES)
+            buf = buf + get_int_as_bytes(tzoffset, TZOFFSET_SIZE_BYTES)
+            buf = buf + get_int_as_bytes(tzindex, TZINDEX_SIZE_BYTES)
+
+        return buf
```

Comment:
Well, since it works with class internals (like `self._datetime.value`), it's better be a part of the class. And I'm not so sure `value` should be exposed as a property.

Reply:
Is it not true `self._datetime.value // NSEC_IN_SEC` == `Int(self.timestamp)` ?

Reply:
Well, until there is no precision loss on conversion from `float` to `int`.

Reply:
https://github.com/pandas-dev/pandas/blob/6dc589b957e494750351e8572b45b8fa12c27d01/pandas/_libs/tslibs/timestamps.pyx#L1139-L1154

long(self.timestamp) ?

Reply:
There is no separate `long` in Python 3, `int` is already an integer object with arbitrary length. I mean that it's not good to rely on float precision if we can rely on integer.

Reply:
But then it will be possible not to violate encapsulation. This should not be a critical problem to cast a float value into a int64-value especially because we have this.sec and this.nsec values.

Reply:
I'm not quite sure, but it looks like python will auto-floor small approximations in int() case.

https://docs.python.org/3/tutorial/floatingpoint.html

Reply:
But then it will be possible not to violate encapsulation.

We don't since it is a class method and, to be honest, there is no any reasons to extract it. If we want to extract it, it's better to expose value attribute and use it rather that hope that float won't lose any integer parts.

Reply:
Exposed `value` attribute (since it could be useful), but do not extracted `msgpack_encode` to the outside.

Reply:
> We don't since it is a class method and, to be honest, there are no any reasons to extract it.

Single-responsibility principle.

Reply:
Well, I think we may say that the main responsibility of Datetime is to be encoded and decoded.

Reply:
And decoding is already a class method (as a constructor).

Reply:
> Well, I think we may say that the main responsibility of Datetime is to be encoded and decoded.

At this stage the class looks like `DTO` and serialization/deserialization methods is ok. But I guess (I did not look at an inteval pull request) that `AddInterval`/`SubInterval`/`SubDatetime` methods will be added in the future. After that, this class will no longer look like a DTO and serialization/deserialization methos will look strange.

In fact, the class does not require the decode method in the current implementation. There is no need for it.

Reply:
`AddInterval`/`SubInterval`/`SubDatetime`

They are implemented as operators overload.

Reply:
> In fact, the class does not require the decode method in the current implementation. 

Well, I wouldn't be too sure about this. For example, if decode would be external, it would be bad for performance because we would need to create excessive intermediate `pandas.Timestamp` to transform epoch to year/month/day/etc. Moreover, as far as I remember, pandas and Tarantool treat epoch + tz with different approaches (and that's the reason `decode` do `.replace(tzinfo=pytz.UTC).tz_convert(tzinfo)` and `__init__` do `.replace(tzinfo=tzinfo)`), so may need some workarounds.